onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -0,0 +1,476 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <core/common/logging/logging.h>
+
+#include "core/common/safeint.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_lib/nnapi_implementation.h"
+#include "helper.h"
+#include "model_builder.h"
+#include "node_attr_helper.h"
+
+namespace onnxruntime {
+namespace nnapi {
+
+using namespace android::nn::wrapper;
+using std::vector;
+
+const float* GetTensorFloatDataA(const ONNX_NAMESPACE::TensorProto& tensor) {
+  return tensor.float_data().empty()
+             ? reinterpret_cast<const float*>(tensor.raw_data().data())
+             : tensor.float_data().data();
+}
+
+ModelBuilder::ModelBuilder(ONNX_NAMESPACE::ModelProto& model_proto)
+    : nnapi_(NnApiImplementation()), model_proto_(model_proto) {
+  GetAllInitializers();
+  op_builders_ = CreateOpBuilders();
+}
+
+int32_t ModelBuilder::GetAndroidSdkVer() const {
+  return nnapi_ ? nnapi_->android_sdk_version : 0;
+}
+
+bool ModelBuilder::IsNodeSupported(
+    const ONNX_NAMESPACE::NodeProto& node) {
+  if (auto* opBuilder = GetOpBuilder(node)) {
+    return opBuilder->IsOpSupported(*this, node);
+  } else {
+    return false;
+  }
+}
+
+bool IsValidSupportedNodesVec(const std::vector<int>& supported_node_vec,
+                              const ONNX_NAMESPACE::ModelProto& model_proto) {
+  if (!supported_node_vec.empty()) {
+    if (supported_node_vec.size() == 1) {
+      const auto& node = model_proto.graph().node(supported_node_vec[0]);
+      const auto& op = node.op_type();
+      // It is not worth it to perform a single Reshape/Dropout/Identity operator
+      // which is only copying the data in NNAPI
+      // If this is the case, let it fall back
+      if (op == "Reshape" ||
+          op == "Dropout" ||
+          op == "Identity") {
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}  // namespace nnapi
+
+std::vector<std::vector<int>> ModelBuilder::GetSupportedNodes() {
+  std::vector<std::vector<int>> supported_node_vecs;
+  int32_t android_sdk_ver = nnapi_ ? nnapi_->android_sdk_version : 0;
+#ifdef __ANDROID__
+  if (android_sdk_ver < 27) {
+    LOGS_DEFAULT(VERBOSE) << "Android API level "
+                          << android_sdk_ver
+                          << " is lower than 27";
+    return supported_node_vecs;
+  }
+#endif
+
+  std::vector<int> supported_node_vec;
+  for (int i = 0; i < model_proto_.graph().node_size(); i++) {
+    const auto& node(model_proto_.graph().node(i));
+    bool supported = IsNodeSupported(node);
+    LOGS_DEFAULT(VERBOSE) << "Operator type: [" << node.op_type()
+                          << "] index: [" << i
+                          << "] name: [" << node.name()
+                          << "] supported: [" << supported
+                          << "]";
+    if (supported) {
+      supported_node_vec.push_back(i);
+    } else {
```

Comment:
Wondering what joins the two worlds of a node index based on the model proto, and the NodeIndex for a Node in an onnxruntime Graph. Not sure those always match.

Reply:
This will be resolve once we resolve TODO (switch input from ModelProto to Graph)